Add GPU decoder for link based ZS.

fweig opened the pull request:
This PR adds a GPU decoder for the new TPC link-based ZS.

@davidrohr

fweig committed:
GPU/TPCClusterFinder: Add GPU decoder for link based ZS.

davidrohr commented:
Some comments already, some more will come:
- If `tpczslinkmapping.dump` is not present in the standalone version, it just segfaults. Instead it should print an error message that the mapping file is missing.
- compiler warnings
```
2022-05-17@00:08:22:DEBUG:O2:O2:0: /home/qon/alice/sw/SOURCES/O2/dev/0/GPU/GPUTracking/TPCClusterFinder/GPUTPCCFDecodeZS.cxx(271): warning #68-D: integer conversion resulted in a change of sign
2022-05-17@00:08:22:DEBUG:O2:O2:0: 
2022-05-17@00:08:22:DEBUG:O2:O2:0: /home/qon/alice/sw/SOURCES/O2/dev/0/GPU/GPUTracking/TPCClusterFinder/GPUTPCCFDecodeZS.cxx(271): warning #68-D: integer conversion resulted in a change of sign
2022-05-17@00:08:22:DEBUG:O2:O2:0: 
2022-05-17@00:08:22:DEBUG:O2:O2:0: /home/qon/alice/sw/SOURCES/O2/dev/0/GPU/GPUTracking/TPCClusterFinder/GPUTPCCFDecodeZS.cxx(212): warning #177-D: variable "positions" was declared but never referenced
```

davidrohr commented:
```
/home/qon/alice/O2/GPU/GPUTracking/TPCClusterFinder/GPUTPCCFDecodeZS.cxx: In static member function ‘static void o2::gpu::GPUTPCCFDecodeZSLink::DecodeTBSingleThread(o2::gpu::GPUTPCCFDecodeZSLink::processorType&, const unsigned char*, int, const uint32_t*, int, int, int, size_t)’:
/home/qon/alice/O2/GPU/GPUTracking/TPCClusterFinder/GPUTPCCFDecodeZS.cxx:317:19: warning: comparison of integer expressions of different signedness: ‘unsigned int’ and ‘int’ [-Wsign-compare]
  317 |   while (posXbits < nAdc) {
      |          ~~~~~~~~~^~~~~~
```

fweig commented:
Updated the PR to address your remarks, @davidrohr:
- Fixed compiler warnings
- An error is now raised if mapping data is missing when running with ZS
- Fixed build failure with opencl

davidrohr commented on GPU/GPUTracking/Global/GPUChainTrackingClusterizer.cxx:
```diff
@@ -529,11 +530,20 @@ int GPUChainTracking::RunTPCClusterizer(bool synchronizeOutput)
+
+          switch (mCFContext->zsVersion) {
+            default:
+              GPUFatal("Data with invalid TPC ZS mode (%d) received", mCFContext->zsVersion);
```
could you add a break for default, just for consistency, despite the GPUFatal

davidrohr commented on GPU/GPUTracking/Base/GPUReconstructionConvert.cxx:
```diff
@@ -569,7 +569,8 @@ unsigned int zsEncoderImprovedLinkBased::encodeSequence(std::vector<o2::tpc::Dig
   tbHdr->syncOffsetBC = 0;
   tbHdr->syncOffsetCRUCyclesOrLink = link;
-  hdr->nTimeBins = tmpBuffer[k].getTimeStamp() - firstTimebinInPage + 1;
+  // FIXME: Encoder attempts to create pages with 256 timebins, causing this counter to overflow
```
You can remove this, a fix is in #8866

davidrohr commented on GPU/GPUTracking/Base/GPUReconstructionConvert.cxx:
```diff
@@ -457,7 +457,7 @@ void zsEncoderRow::decodePage(std::vector<o2::tpc::Digit>& outputBuffer, const z
 #ifdef GPUCA_O2_LIB
 struct zsEncoderImprovedLinkBased : public zsEncoder {
-  static constexpr bool TIGHTLY_PACKED = false;
+  static constexpr bool TIGHTLY_PACKED = true;
```
The improved link-based for now will only support TIGHTLY_PACKED = false.
Could you adapt the decoder to support both, and set this back to false?

davidrohr reviewed (changes requested):
@fweig 

thx, this looks quite good in general, though I haven't checked yet in too much detail. Some more comments:
: could you fix:
```
========== List of errors found ==========
/sw/SOURCES/O2/8843-slc8_x86-64/0//GPU/GPUTracking/DataTypes/TPCZSLinkMapping.h:1:1: error: missing or malformed copyright notice
/sw/SOURCES/O2/8843-slc8_x86-64/0//GPU/GPUTracking/DataTypes/TPCZSLinkMapping.cxx:1:1: error: missing or malformed copyright notice
```

And there is one more compiler warning:
```
/home/qon/alice/O2/GPU/GPUTracking/TPCClusterFinder/GPUTPCCFDecodeZS.cxx: In static member function ‘static void o2::gpu::GPUTPCCFDecodeZSLink::Thread(int, int, int, int, o2::gpu::GPUTPCCFDecodeZSLink::GPUSharedMemory&, o2::gpu::GPUTPCCFDecodeZSLink::processorType&, Args ...) [with int iKernel = 0; Args = {int}; o2::gpu::GPUTPCCFDecodeZSLink::processorType = o2::gpu::GPUTPCClusterFinder]’:
/home/qon/alice/O2/GPU/GPUTracking/TPCClusterFinder/GPUTPCCFDecodeZS.cxx:268:35: warning: comparison of integer expressions of different signedness: ‘int’ and ‘unsigned int’ [-Wsign-compare]
  268 |           for (int a = iThread; a < nAdc; a += nThreads) {
      |  
```

davidrohr commented:
@fweig : Another comment, you must adapt slightly to #8877 . The required changes are (as patch file):
```
diff --git a/GPU/GPUTracking/TPCClusterFinder/GPUTPCCFDecodeZS.cxx b/GPU/GPUTracking/TPCClusterFinder/GPUTPCCFDecodeZS.cxx
index 71c737322ff..3bed5afe1db 100644
--- a/GPU/GPUTracking/TPCClusterFinder/GPUTPCCFDecodeZS.cxx
+++ b/GPU/GPUTracking/TPCClusterFinder/GPUTPCCFDecodeZS.cxx
@@ -246,7 +246,7 @@ GPUdii() void GPUTPCCFDecodeZSLink::Thread<0>(int nBlocks, int nThreads, int iBl
       page += decHdr->firstZSDataOffset * 16;
 
       assert(decHdr->version == 3);
-      assert(decHdr->magicWord == 0x1234);
+      assert(decHdr->magicWord == o2::tpc::zerosupp_link_based::CommonHeader::MagicWordLinkZSMetaHeader);
 
       for (unsigned int t = 0; t < decHdr->nTimebinHeaders; t++) {
         const zerosupp_link_based::CommonHeader* tbHdr = (const zerosupp_link_based::CommonHeader*)page;
@@ -283,7 +283,7 @@ GPUdii() void GPUTPCCFDecodeZSLink::Thread<0>(int nBlocks, int nThreads, int iBl
           channelMask,
           timeBin,
           decHdr->cruID,
-          tbHdr->syncOffsetCRUCyclesOrLink,
+          tbHdr->fecInPartition,
           pageDigitOffset);
 #else // CPU
         DecodeTBSingleThread(
@@ -293,7 +293,7 @@ GPUdii() void GPUTPCCFDecodeZSLink::Thread<0>(int nBlocks, int nThreads, int iBl
           channelMask,
           timeBin,
           decHdr->cruID,
-          tbHdr->syncOffsetCRUCyclesOrLink,
+          tbHdr->fecInPartition,
           pageDigitOffset);
 #endif
         pageDigitOffset += nAdc;
```

fweig commented:
@davidrohr Thanks, will do! Hope I can do the rebase / update tomorrow.

fweig commented:
Updated the PR again. Decoder now supports tightly packed and non tightly packed data. Should've addressed your other points as well and did some (minor) refactor. Note that without the tightly packed pages the performance gains in the decoder are gone now.

davidrohr reviewed (commented):
Thx @fweig : This looks pretty good. Just one more request. Could you move the tightly packed constexpr value to the ZeroSuppression.h header and use that flag in the encoder and the decoder, and second the coding guidelines define that constexpr variables should be all capital.

fweig commented:
> Thx @fweig : This looks pretty good. Just one more request. Could you move the tightly packed constexpr value to the ZeroSuppression.h header and use that flag in the encoder and the decoder, and second the coding guidelines define that constexpr variables should be all capital.

Fixed. Renamed the other constants as well.
